pkg/didcomm/protocol/didexchange/service_test.go

```diff
@@ -478,30 +486,354 @@ func (p *mockProvider) OutboundTransport() transport.OutboundTransport {
 	return didcomm.NewMockOutboundTransport(successResponse)
 }
 
-//TODO: Use in memory store rather than level DB " https://github.com/hyperledger/aries-framework-go/issues/202
-func store(t testing.TB) (store storage.Store, cleanup func()) {
-	path, cleanup := tempDir(t)
-	prov, err := leveldb.NewProvider(path)
-	require.NoError(t, err)
+func store(t testing.TB) (storage.Store, func()) {
+	prov := mockstorage.NewMockStoreProvider()
 	dbstore, err := prov.GetStoreHandle()
 	require.NoError(t, err)
-	return dbstore, cleanup
+	return dbstore, func() {
+		err := prov.Close()
+		require.NoError(t, err)
+	}
+}
+
+func randomString() string {
+	u := uuid.New()
+	return u.String()
+}
+
+type payload struct {
+	ID string `json:"@id"`
+}
+
+func TestService_Events(t *testing.T) {
+	dbstore, cleanup := store(t)
+	defer cleanup()
+
+	svc := New(dbstore, &mockProvider{})
+
+	startConsumer(t, svc)
+
+	validateSuccessCase(t, svc)
+
+	validateUserError(t, svc)
+
+	validateStoreError(t, svc)
+
+	validateHandleError(t, svc)
+
+	validateStoreDataCorruptionError(t, svc)
 }
 
-func tempDir(t testing.TB) (string, func()) {
-	dbPath, err := ioutil.TempDir("", "db")
-	if err != nil {
-		t.Fatalf("Failed to create leveldb directory: %s", err)
+func startConsumer(t *testing.T, svc *Service) {
+	invitationChannel := make(chan event.DIDCommEvent)
+	err := svc.Register(ConnectionInvite, invitationChannel)
+	require.NoError(t, err)
+
+	err = svc.Register(ConnectionInvite, make(chan event.DIDCommEvent))
+	require.NoError(t, err)
+
+	err = svc.Register("invalid-type", make(chan event.DIDCommEvent))
+	require.Error(t, err)
+	require.Contains(t, err.Error(), "event registration not supported for the message type invalid-type")
+
+	requestChannel := make(chan event.DIDCommEvent)
+	err = svc.Register(ConnectionRequest, requestChannel)
+	require.NoError(t, err)
+
+	responseChannel := make(chan event.DIDCommEvent)
+	err = svc.Register(ConnectionResponse, responseChannel)
+	require.NoError(t, err)
+
+	ackChannel := make(chan event.DIDCommEvent)
+	err = svc.Register(ConnectionAck, ackChannel)
+	require.NoError(t, err)
+
+	pl := payload{}
+
+	go func() {
+		for {
+			select {
+			// receive the event on ConnectionRequest message type
+			case e := <-invitationChannel:
+				require.NotEmpty(t, e.Message)
+				require.Equal(t, ConnectionInvite, e.Message.Type)
+
+				// invoke callback
+				err = e.Callback(event.DIDCommCallback{Err: nil})
```

Comment:
When Callback is invoked, processCallback will be invoked and handle(...) will now run in this Go routine (i.e. the client's Go routine). This Go routine will be blocked until the flow has completed which may result in deadlock if the client needs to handle more events.

Reply:
@bstasyszyn Very valid point. I was thinking to reduce the work for consumers to call the go-channel as they had to map the correlation keys.

For this, may be we can start a go routine or invoke channel inside call back function. So that it can be returned the control back immediately. If this is not a blocker for now, can we tackle this issue in next PRs/issue ? I can create the issue for this.

Reply:
[Wrapped](https://github.com/hyperledger/aries-framework-go/compare/9e40655340de2e0f5c4fc9f2074c4f8cd3e1126d..686bbecbf5c870012cfbab30d759addb432f52e9#diff-f761c91e753eba1c42a68076df06ce12R207) internal go channel inside the callback function for non-blocking callback from the consumer.